src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs

```diff
@@ -341,7 +371,7 @@ public override int GetHashCode()
         /// <summary>
         /// RID -> RID compatibility key
         /// </summary>
-        private sealed class RuntimeCompatKey : IEquatable<RuntimeCompatKey>
+        private readonly struct RuntimeCompatKey : IEquatable<RuntimeCompatKey>
```

Comment:
Convert this private collection key to a struct to reduce heap allocations.

Reply:
Today, I learned about the concept of a readonly struct. According to the documentation, a `readonly struct` can undergo [boxing and unboxing conversions](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/struct#conversions) when it implements an interface. In this case, the `RuntimeCompatKey` is used as a key in a dictionary, and it implements the `IEquatable` interface. Therefore, I would anticipate that there might be some boxing operations involved when comparing keys for equality. Is it acceptable for these allocations to occur due to boxing?

Reply:
Structs are only boxed when they're called through the interface.

```c#
int i = 1;
i.CompareTo(2); // call (non boxing and cheap)
IComparable j = i; // box
j.CompareTo(2); // callvirt and boxed argument (expensive)
```

Here's the MSIL for the above code.

```
// i.CompareTo(2)
        IL_0002: ldloca.s 0
        IL_0004: ldc.i4.2
        IL_0005: call instance int32 [mscorlib]System.Int32::CompareTo(int32)
        IL_000a: pop
// IComparable j = i
        IL_000b: ldloc.0
        IL_000c: box [mscorlib]System.Int32
// j.CompareTo(2)
        IL_0011: ldc.i4.2
        IL_0012: box [mscorlib]System.Int32
        IL_0017: callvirt instance int32 [mscorlib]System.IComparable::CompareTo(object)
        IL_001c: pop
        IL_001d: ret
```

In the case of the dictionary, if `TKey` is a struct, the JIT generates non-boxing calls to the `Equals` and `GetHashCode` overridden from `object`.

We could probably remove the `IEquatable<>` interfaces from this type and there'd be no difference in behaviour.